File: Trie.cpp
#include<bits/stdc++.h>
using namespace std;
const int alph =26;
const int maxn=100000;
struct Trie{
    int trie[maxn][alph], nc, hot_nodes[maxn];
    int max_depth;
    void init(){
        for(int i=0;i<maxn;i++){
            for(int j=0;j<alph;j++){
                trie[i][j]=-1;
            }
        }
        nc=0;
        max_depth=0;
        for(int i=0;i<maxn;i++){
            hot_nodes[i]=0;
        }
    }
    void insert(string s){
        int cur=0;
        for(int i=0;i<s.length();i++){
            int c=s[i]-'a';
            if(trie[cur][c]==-1){
                trie[cur][c]=++nc;
            }
            cur=trie[cur][c];
        }
        hot_nodes[cur]++;
    }
    bool exact_check(string s){
        int cur=0;
        for(int i=0;i<s.length();i++){
            int c=s[i]-'a';
            if(trie[cur][c]==-1){
                return false;
            }
            cur=trie[cur][c];
        }
        if(hot_nodes[cur]>0){
            return true;
        }
        else{
            return false;
        }
    }
    bool prefix_check(string s){
        int cur=0;
        for(int i=0;i<s.length();i++){
            int c=s[i]-'a';
            if(trie[cur][c]==-1){
                return false;
            }
            cur=trie[cur][c];
        }
        return true;
    }
    int longest_prefix(){
        dfs(0,0);
    }
    int dfs(int n,int depth){
        int sum=hot_nodes[n];
        for(int i=0;i<26;i++){
            if(trie[n][i]!=-1){
                sum+=dfs(trie[n][i],depth+1);
            }
        }
        if(sum>=2&&depth>max_depth){
        max_depth=depth;
        }
        return sum;
    }
}
int main(){
    
}

File: trie.py
class TRIE:
    def __init__(self, n) -> None:
        self.alph =26
        self.maxn=100000
        self.trie=[[-1]*self.alph]*self.maxn
        self.max_depth=0
        self.nc=0
        self.hot_nodes=[1]*maxn
    
    def insert(self, s):
        cur=0
        for i in range(len(s)):
            c=ord(s[i])-ord('a')
            if self.trie[cur][c] == -1:
                self.nc+=1
                self.trie[cur][c]=self.nc
            cur=self.trie[cur][c]
        self.hot_nodes[cur]+=1

    def exact_check(self, x):
        cur=0
        for i in range(len(x)):
            c=ord(s[i])-ord('a')
            if self.trie[cur][c] == -1:
                return False
            cur=self.trie[cur][c]
        if self.hot_nodes[cur] >0 :
            return True
        else:
            return False
    
    def prefix_check(self, x):
        cur=0
        for i in range(len(x)):
            c=ord(s[i])-ord('a')
            if self.trie[cur][c] == -1:
                return False
            cur=self.trie[cur][c]
        return True
